## Huang-yuting-final-project2

### group project: Yuting Huang, Jennifer Shiyue Zhang, Chung Kuang Lin

In [1]:
import pandas as pd
import matplotlib.pyplot as plt
import bqplot
import numpy as np
import ipywidgets
import traitlets

In [2]:
nutrition = pd.read_csv('Nutrition__Physical_Activity__and_Obesity_-_Behavioral_Risk_Factor_Surveillance_System.csv')
nutrition

YearStart  YearEnd LocationAbbr LocationDesc  \
0           2014     2014           GU         Guam   
1           2013     2013           US     National   
2           2013     2013           US     National   
3           2015     2015           US     National   
4           2015     2015           GU         Guam   
...          ...      ...          ...          ...   
80924       2019     2019           MI     Michigan   
80925       2019     2019           WI    Wisconsin   
80926       2019     2019           NE     Nebraska   
80927       2019     2019           TX        Texas   
80928       2019     2019           MO     Missouri   

                                       Datasource                    Class  \
0      Behavioral Risk Factor Surveillance System  Obesity / Weight Status   
1      Behavioral Risk Factor Surveillance System  Obesity / Weight Status   
2      Behavioral Risk Factor Surveillance System  Obesity / Weight Status   
3      Behavioral Risk Factor Surveillance System        Physical Activity   
4      Behavioral Risk Factor Surveillance System        Physical Activity   
...                                           ...                      ...   
80924  Behavioral Risk Factor Surveillance System    Fruits and Vegetables   
80925  Behavioral Risk Factor Surveillance System    Fruits and Vegetables   
80926  Behavioral Risk Factor Surveillance System    Fruits and Vegetables   
80927  Behavioral Risk Factor Surveillance System    Fruits and Vegetables   
80928  Behavioral Risk Factor Surveillance System    Fruits and Vegetables   

                                  Topic  \
0               Obesity / Weight Status   
1               Obesity / Weight Status   
2               Obesity / Weight Status   
3          Physical Activity - Behavior   
4          Physical Activity - Behavior   
...                                 ...   
80924  Fruits and Vegetables - Behavior   
80925  Fruits and Vegetables - Behavior   
80926  Fruits and Vegetables - Behavior   
80927  Fruits and Vegetables - Behavior   
80928  Fruits and Vegetables - Behavior   

                                                Question  Data_Value_Unit  \
0      Percent of adults aged 18 years and older who ...              NaN   
1      Percent of adults aged 18 years and older who ...              NaN   
2      Percent of adults aged 18 years and older who ...              NaN   
3      Percent of adults who achieve at least 300 min...              NaN   
4      Percent of adults who achieve at least 150 min...              NaN   
...                                                  ...              ...   
80924  Percent of adults who report consuming fruit l...              NaN   
80925  Percent of adults who report consuming vegetab...              NaN   
80926  Percent of adults who report consuming fruit l...              NaN   
80927  Percent of adults who report consuming fruit l...              NaN   
80928  Percent of adults who report consuming fruit l...              NaN   

      Data_Value_Type  ...                               GeoLocation  ClassID  \
0               Value  ...                   (13.444304, 144.793731)      OWS   
1               Value  ...                                       NaN      OWS   
2               Value  ...                                       NaN      OWS   
3               Value  ...                                       NaN       PA   
4               Value  ...                   (13.444304, 144.793731)       PA   
...               ...  ...                                       ...      ...   
80924           Value  ...    (44.6613195430005, -84.71439026999968)       FV   
80925           Value  ...   (44.39319117400049, -89.81637074199966)       FV   
80926           Value  ...    (41.6410409880005, -99.36572062299967)       FV   
80927           Value  ...  (31.827240407000488, -99.42677020599967)       FV   
80928           Value  ...  (38.635790776000476, -92.56630005299968)       F

In [3]:
Location = nutrition['LocationAbbr'].unique()
Location

array(['GU', 'US', 'WY', 'DC', 'PR', 'AL', 'RI', 'NJ', 'WA', 'MI', 'VA',
       'CA', 'UT', 'NY', 'MA', 'DE', 'AR', 'IL', 'NH', 'NM', 'MD', 'HI',
       'LA', 'TX', 'KY', 'SD', 'WV', 'CO', 'OK', 'MS', 'OR', 'WI', 'KS',
       'FL', 'ID', 'ME', 'AZ', 'VI', 'MT', 'MN', 'GA', 'NC', 'PA', 'ND',
       'SC', 'NE', 'TN', 'MO', 'NV', 'IA', 'IN', 'OH', 'AK', 'VT', 'CT'],
      dtype=object)

In [4]:
nutrition.dtypes

YearStart                       int64
YearEnd                         int64
LocationAbbr                   object
LocationDesc                   object
Datasource                     object
Class                          object
Topic                          object
Question                       object
Data_Value_Unit               float64
Data_Value_Type                object
Data_Value                    float64
Data_Value_Alt                float64
Data_Value_Footnote_Symbol     object
Data_Value_Footnote            object
Low_Confidence_Limit          float64
High_Confidence_Limit         float64
Sample_Size                   float64
Total                          object
Age(years)                     object
Education                      object
Gender                         object
Income                         object
Race/Ethnicity                 object
GeoLocation                    object
ClassID                        object
TopicID                        object
QuestionID  

In [5]:
#nutrition['Sample_Size'] = nutrition['Sample_Size'].str.replace(",", "").astype(float)

In [6]:
# Replace the "Age (years)" value in StratificationCategory1 column into the "Age(years)" as the column with same name
# This will help the dashbroad to work properly.
nutrition["StratificationCategory1"] = nutrition["StratificationCategory1"].str.replace(" ", "")

In [7]:
nutrition["StratificationCategory1"].values

array(['Education', 'Income', 'Income', ..., 'Age(years)', 'Education',
       'Income'], dtype=object)

In [8]:
nutrition["Sample_Size"]

0          842.0
1        62562.0
2        60069.0
3        30904.0
4          125.0
          ...   
80924     1440.0
80925     1027.0
80926      874.0
80927     2897.0
80928     1267.0
Name: Sample_Size, Length: 80929, dtype: float64

In [9]:
# Use pivot table to find the relationship between 'LocationAbbr' and 'StratificationCategory1'
table = pd.pivot_table(nutrition, values = 'Sample_Size', index = ['LocationAbbr'],
                      columns = 'StratificationCategory1', aggfunc = np.mean)
table.fillna(0)

StratificationCategory1    Age(years)      Education         Gender  \
LocationAbbr                                                          
AK                         549.567901     822.175926    1649.629630   
AL                        1113.555556    1668.245370    3342.351852   
AR                         793.546296    1187.615741    2381.509259   
AZ                        1417.373457    2120.476852    4252.796296   
CA                        1780.802469    2663.490741    5342.407407   
CO                        1784.950617    2672.754630    5357.490741   
CT                        1414.595679    2116.995370    4245.703704   
DC                         609.086420     911.620370    1829.379630   
DE                         676.179012    1011.837963    2030.027778   
FL                        2752.725309    4115.092593    8260.638889   
GA                        1085.330247    1623.921296    3257.638889   
GU                         295.977477     443.054054     887.932432   
HI                        1201.033951    1799.745370    3604.194444   
IA                        1238.064815    1854.814815    3714.916667   
ID                         839.743827    1256.750000    2520.453704   
IL                         838.151235    1256.185185    2514.777778   
IN                        1469.422840    2198.550926    4409.416667   
KS                        2676.580247    4009.240741    8030.546296   
KY                        1401.756173    2100.134259    4207.175926   
LA                         923.259259    1383.504630    2770.851852   
MA                        1774.663580    2655.111111    5332.037037   
MD                        2121.675926    3174.129630    6368.250000   
ME                        1604.083333    2401.597222    4813.277778   
MI                        1645.083333    2464.268519    4937.916667   
MN                        2424.925926    3631.356481    7277.074074   
MO                        1118.922840    1676.046296    3358.064815   
MS                         966.879630    1447.587963    2901.953704   
MT                        1144.978395    1715.226852    3435.851852   
NC                        1089.000000    1631.541667    3269.777778   
ND                         934.561728    1399.648148    2804.490741   
NE                        2771.033951    4151.814815    8315.407407   
NH                         959.783951    1437.277778    2880.666667   
NJ                        1802.885185    2698.250000    5415.744444   
NM                        1151.484568    1724.814815    3456.444444   
NV                         600.336420     899.458333    1803.185185   
NY                        2038.126543    3043.300926    6108.768519   
OH                        1876.456790    2810.060185    5632.333333   
OK                        1094.728395    1639.949074    3284.703704   
OR                         867.243827    1297.134259    2602.398148   
PA                        1339.435185    2007.231481    4023.518519   
PR                         860.991453    1288.147436    2582.858974   
RI                         910.324074    1362.824074    2732.166667   
SC                        1607.160494    2406.370370    4823.546296   
SD                        1115.219136    1670.689815    3346.527778   
TN                         882.145062    1320.310185    2647.435185   
TX                        1897.941358    2838.800926    5694.277778   
US                       68862.296296  103091.412037  206668.564815   
UT                        1823.935185    2730.856481    5472.777778   
VA                        1359.398148    2034.597222    4080.601852   
VI                         203.944444     300.500000     611.833333   
VT                        1005.037037    1504.907407    3015.240741   
WA                        2073.858025    3105.199074    6224.083333   
WI                         860.780864    1288.972222    2583.138889   
WV                         892.895062    1337.643519    2679.490741   
WY                      

In [10]:
table.values

array([[5.49567901e+02, 8.22175926e+02, 1.64962963e+03, 4.71386243e+02,
        6.32594796e+02, 3.29970370e+03],
       [1.11355556e+03, 1.66824537e+03, 3.34235185e+03, 9.55018519e+02,
        1.33624621e+03, 6.68512963e+03],
       [7.93546296e+02, 1.18761574e+03, 2.38150926e+03, 6.80589947e+02,
        9.11727273e+02, 4.76412963e+03],
       [1.41737346e+03, 2.12047685e+03, 4.25279630e+03, 1.21555291e+03,
        1.24926462e+03, 8.50887037e+03],
       [1.78080247e+03, 2.66349074e+03, 5.34240741e+03, 1.52644974e+03,
        1.55311202e+03, 1.06851481e+04],
       [1.78495062e+03, 2.67275463e+03, 5.35749074e+03, 1.53089418e+03,
        1.55780939e+03, 1.07162593e+04],
       [1.41459568e+03, 2.11699537e+03, 4.24570370e+03, 1.21341534e+03,
        1.36955963e+03, 8.49390741e+03],
       [6.09086420e+02, 9.11620370e+02, 1.82937963e+03, 5.22809524e+02,
        6.87826715e+02, 3.65966667e+03],
       [6.76179012e+02, 1.01183796e+03, 2.03002778e+03, 5.80007937e+02,
        8.21260700e+02, 

In [11]:
Stratification = table.columns.to_numpy()
location = table.index.to_numpy()

In [12]:
Stratification

array(['Age(years)', 'Education', 'Gender', 'Income', 'Race/Ethnicity',
       'Total'], dtype=object)

In [13]:
location

array(['AK', 'AL', 'AR', 'AZ', 'CA', 'CO', 'CT', 'DC', 'DE', 'FL', 'GA',
       'GU', 'HI', 'IA', 'ID', 'IL', 'IN', 'KS', 'KY', 'LA', 'MA', 'MD',
       'ME', 'MI', 'MN', 'MO', 'MS', 'MT', 'NC', 'ND', 'NE', 'NH', 'NJ',
       'NM', 'NV', 'NY', 'OH', 'OK', 'OR', 'PA', 'PR', 'RI', 'SC', 'SD',
       'TN', 'TX', 'US', 'UT', 'VA', 'VI', 'VT', 'WA', 'WI', 'WV', 'WY'],
      dtype=object)

In [14]:
# I tried to change their datatype to lists because I did in my homework 6.
# However, I found there is no differences between an array and a list if I put them into the heatmap.

#location_array = [i for i in location]
#location_array
#Strat_array = [i for i in Stratification]
#Strat_array

In [15]:
table.values


array([[5.49567901e+02, 8.22175926e+02, 1.64962963e+03, 4.71386243e+02,
        6.32594796e+02, 3.29970370e+03],
       [1.11355556e+03, 1.66824537e+03, 3.34235185e+03, 9.55018519e+02,
        1.33624621e+03, 6.68512963e+03],
       [7.93546296e+02, 1.18761574e+03, 2.38150926e+03, 6.80589947e+02,
        9.11727273e+02, 4.76412963e+03],
       [1.41737346e+03, 2.12047685e+03, 4.25279630e+03, 1.21555291e+03,
        1.24926462e+03, 8.50887037e+03],
       [1.78080247e+03, 2.66349074e+03, 5.34240741e+03, 1.52644974e+03,
        1.55311202e+03, 1.06851481e+04],
       [1.78495062e+03, 2.67275463e+03, 5.35749074e+03, 1.53089418e+03,
        1.55780939e+03, 1.07162593e+04],
       [1.41459568e+03, 2.11699537e+03, 4.24570370e+03, 1.21341534e+03,
        1.36955963e+03, 8.49390741e+03],
       [6.09086420e+02, 9.11620370e+02, 1.82937963e+03, 5.22809524e+02,
        6.87826715e+02, 3.65966667e+03],
       [6.76179012e+02, 1.01183796e+03, 2.03002778e+03, 5.80007937e+02,
        8.21260700e+02, 

In [16]:
# Observe each variable's shape is important. 
# I met lots of errors because the selection is over the row boundary.
Stratification.shape

(6,)

In [17]:
location.shape

(55,)

In [18]:
table.values.shape

(55, 6)

In [19]:
#Reference: https://stackoverflow.com/questions/37144008/change-size-of-fig-in-bqplot (change height)

In [20]:
##GRID HEAT MAP##

# 2. scales
col_sc = bqplot.ColorScale(scheme='Purples')
x_sc = bqplot.OrdinalScale()
y_sc = bqplot.OrdinalScale()

# 3. axis
col_ax = bqplot.ColorAxis(scale=col_sc, orientation='horizontal', side='top', label = 'log10 Color Scale: ')
x_ax = bqplot.Axis(scale=x_sc, label='Stratification',tick_rotate=320, tick_style={'font-size': '14px'}, label_offset ='50px')

y_ax = bqplot.Axis(scale=y_sc, orientation='vertical', label='Location',tick_offset = '50px',tick_style ={'font-size':'14px'}, label_offset ='50px')

# 4. marks
heat_map = bqplot.GridHeatMap(color = np.log10(table.values),
                             row = location,
                             column = Stratification, 
                             scales = {'color':col_sc,'row':y_sc, 'column':x_sc},
                             interactions={'click':'select'},
                             selected_style={'fill':'pink'})

fig = bqplot.Figure(marks=[heat_map], axes=[col_ax,x_ax, y_ax])
fig.layout.height = '800px'

In [21]:
fig

Figure(axes=[ColorAxis(label='log10 Color Scale: ', scale=ColorScale(scheme='Purples'), side='top'), Axis(labe…

In [22]:
myLabel = ipywidgets.Label()

In [23]:
print(table.index)
print(table.columns)

Index(['AK', 'AL', 'AR', 'AZ', 'CA', 'CO', 'CT', 'DC', 'DE', 'FL', 'GA', 'GU',
       'HI', 'IA', 'ID', 'IL', 'IN', 'KS', 'KY', 'LA', 'MA', 'MD', 'ME', 'MI',
       'MN', 'MO', 'MS', 'MT', 'NC', 'ND', 'NE', 'NH', 'NJ', 'NM', 'NV', 'NY',
       'OH', 'OK', 'OR', 'PA', 'PR', 'RI', 'SC', 'SD', 'TN', 'TX', 'US', 'UT',
       'VA', 'VI', 'VT', 'WA', 'WI', 'WV', 'WY'],
      dtype='object', name='LocationAbbr')
Index(['Age(years)', 'Education', 'Gender', 'Income', 'Race/Ethnicity',
       'Total'],
      dtype='object', name='StratificationCategory1')


In [24]:
def on_selection(change):
    if len(change['owner'].selected==1): # only allow user to select one grid
        i,j = change['owner'].selected[0]
        v = table.values[i,j]
        #myLabel = myLabel(color = '#ff0000')
        myLabel.value = 'Total sample size for ' +  Stratification[j] +' class '+'in '+ location[i] +' is: ' +str(v) +' ft^2'
        
        new_location = table.index[i]
        new_category = table.columns[j]
        filter = (nutrition["LocationAbbr"] == new_location) & (nutrition["StratificationCategory1"] == new_category) & (nutrition["Question"] == "Percent of adults aged 18 years and older who have an overweight classification")
        subset = nutrition[filter]
        
        strata_type = subset[new_category].values
        percentage = subset["Data_Value"].values
        
        bar.x = strata_type
        bar.y = percentage

In [25]:
heat_map.observe(on_selection,'selected')

In [26]:
fig.layout.min_width='300px'
fig.layout.min_height='300px'
myDashboard = ipywidgets.VBox([myLabel,fig])

In [27]:
myDashboard

In [28]:
# Scales
x_scb = bqplot.OrdinalScale()
y_scb = bqplot.LinearScale()

# Axis
x_axb = bqplot.Axis(label='Stratification Type', scale=x_scb, tick_rotate=330)
y_axb = bqplot.Axis(label="Percentage of adults aged 18+ and classified as obesity", scale=y_scb, orientation='vertical')

i, j = 0, 0
new_location = table.index[i]
new_stratification = table.columns[j]

# mask
mask = nutrition[(nutrition["LocationAbbr"] == new_location) & (nutrition["StratificationCategory1"] == new_stratification) & (nutrition["Question"] == "Percent of adults aged 18 years and older who have an overweight classification")]
strata_type = mask[new_stratification].values
percentage = mask["Data_Value"].values

# Marks
bar = bqplot.Bars(x=strata_type,
                  y=percentage,
                  scales={"x": x_scb, "y": y_scb})
                  
fig_bar = bqplot.Figure(marks=[bar], axes=[x_axb, y_axb], padding_x=0.025, padding_y=0.025)
fig_bar

Figure(axes=[Axis(label='Stratification Type', scale=OrdinalScale(), tick_rotate=330), Axis(label='Percentage …

In [29]:
fig.layout.min_width='500px'
fig.layout.min_height='300px'
fig_bar.layout.min_width = '600px'
fig_bar.layout.minheight = '900px'
new_dashboard = ipywidgets.HBox([myDashboard, fig_bar])

In [101]:
new_dashboard.min_width="900px"

In [102]:
new_dashboard

## Part 2

### One paragraph explaining how to use the dashboard you created, to help someone who is not an expert understand your dataset.

The dataset we select shows US adults' fruits and vegetable diet, physical activity, and weight status from the Behavioral Risk Factor Surveillance System, which includes 80,929 observations. It was retrieved from the Centers for Disease Control and Prevention (2021).

The left heatmap shows the sample size of the U.S. and states. Each cell represents a sample size of either the U.S. or states in terms of various stratifications, namely, age, education, gender, income, and race. The darker color indicates larger sample sizes while the lighter color indicates smaller sample sizes. When clicking on a random cell, the selected cell turns pink. The upper part of the heatmap summarizes the sample information including location and stratification. Accordingly, the right bar chart will show the obesity percentage of 18+ adults in terms of the selected stratification type. For example, when choosing a cell under the Age(years) stratification, the bar chart will show the data on six different levels of Age(years).


### related dataset
1. Name: Table of Overweight and Obesity (BMI)
URL:https://chronicdata.cdc.gov/api/views/fqb7-mgjf/rows.csv?accessType=DOWNLOAD&bom=true&format=true
Why: I think this table could help viewers understand the data of Overweight and Obesity class because it provides BMI distribution cross the U.S. Furthermore, the data is also generated from the same system, Behavioral Risk Factor Surveillance System (BRFSS).

1. Name: Nutrition, Physical Activity, and Obesity - Youth Risk Behavior Surveillance System
URL: https://chronicdata.cdc.gov/api/views/vba9-s8jp/rows.csv?accessType=DOWNLOAD
Why: This is a youth dataset about nutrition, physical activity, and obesity. I think using our dataset which subjects are adults and this youth dataset would help us understand the U.S. citizens' behavior on this topic.

## Additional plot - Yuting

When I saw our dashboard, I found that perhaps building a plot in that people could choose the x and y-axis would be an idea. Therefore, I use the ipywidgets. Dropdown function to let viewers select x and y. I specify the options in x are 'LocationAbbr' and 'Class' because these two variables do not make an error. Moreover, I design the options in y are 'number' data types except 'float64' because 'float64' would make an error. Also, I use the scatter.legend_elements() function to connect scatter and build the legend.

In [103]:
nutrition.dtypes

YearStart                       int64
YearEnd                         int64
LocationAbbr                   object
LocationDesc                   object
Datasource                     object
Class                          object
Topic                          object
Question                       object
Data_Value_Unit               float64
Data_Value_Type                object
Data_Value                    float64
Data_Value_Alt                float64
Data_Value_Footnote_Symbol     object
Data_Value_Footnote            object
Low_Confidence_Limit          float64
High_Confidence_Limit         float64
Sample_Size                   float64
Total                          object
Age(years)                     object
Education                      object
Gender                         object
Income                         object
Race/Ethnicity                 object
GeoLocation                    object
ClassID                        object
TopicID                        object
QuestionID  

In [104]:
type(nutrition)

pandas.core.frame.DataFrame

In [105]:
#x = ipywidgets.Dropdown(options = list(nutrition.select_dtypes(exclude = 'number').columns), description = 'Select a categorical column:')
x = ipywidgets.Dropdown(options = ['LocationAbbr', 'Class'], description = 'Select a categorical column:')
x

Dropdown(description='Select a categorical column:', options=('LocationAbbr', 'Class'), value='LocationAbbr')

In [106]:
x.style.description_width = 'initial'
x

Dropdown(description='Select a categorical column:', options=('LocationAbbr', 'Class'), style=DescriptionStyle…

In [107]:
#y = ipywidgets.Dropdown(options = list(nutrition.select_dtypes(include='number').columns), description = 'Select a numerical column:')
y = ipywidgets.Dropdown(options = list(nutrition.select_dtypes(include='number', exclude = 'float64').columns), description = 'Select a numerical column:')
y

Dropdown(description='Select a numerical column:', options=('YearStart', 'YearEnd', 'LocationID'), value='Year…

In [108]:
y.style.description_width = 'initial'
y

Dropdown(description='Select a numerical column:', options=('YearStart', 'YearEnd', 'LocationID'), style=Descr…

In [109]:
# Reference: https://matplotlib.org/stable/gallery/lines_bars_and_markers/scatter_with_legend.html
def nutrition_xy(x, y):
    with plt.style.context('seaborn'):
        #fig = plt.figure(figsize=(16, 10))
        fig, ax = plt.subplots(figsize=(16, 10))
        scatter = ax.scatter(x = nutrition[x], y = nutrition[y],c =np.log10(nutrition['Sample_Size'])  ,s = 20, cmap = 'Accent')
        legend1 = ax.legend(*scatter.legend_elements(),
                    loc="upper right", title="Sample_Size")
        #plt.scatter(x = nutrition[x], y = nutrition[y],c =np.log10(nutrition['Sample_Size'])  ,s = 20,cmap = 'rainbow')
        plt.xlabel(x.capitalize())
        plt.ylabel(y.capitalize())
        plt.title("%s vs %s"%(x.capitalize(), y.capitalize()))
        plt.show()

In [110]:
out = ipywidgets.interactive_output(nutrition_xy, {'x': x, 'y':y})

In [111]:
ipywidgets.VBox([ipywidgets.HBox([x, y]), out])

### What I learned from my own and other members

I build a pivot table with ‘Sample_Size’, ‘LocationAbbr’, and ‘StratificationCattegory1’, but I would like to include ‘Yearstart’ first in the pivot table because it helps make sense of the plot I made later. However, it is difficult to retrieve two variables from the index since the index.to_numpy() function can only extract one variable. Secondly, I found the data type of variables, such as a list and an array is not important when I put them into a heatmap. The most important thing I learned on my own is shape. Each variable shape decides row and column.


I learned from Chung Kuang is using Boolean to retrieve the data from the nutrition dataset. First, he uses the i and j from the heatmap to connect the bar chart. Then, to find the type of each stratification, he builds a new column to find the value by using the values function instead use the column name, ‘Stratification1’.

I learned from Jennifer Shiyue Zhang that is detailed and clearly describes our works which I think is important. When we enter into a company, we are required to transfer our codes into an easily understood language for those who are not familiar with our job. It helps us to cooperate with other teams in the company.



### 1-2 paragraphs describing your outline for what you will build for Part 3 including how work will be distributed among group members.  Include any drawings/schematics of your proposed website/notebook here as appropriate.

If I would like to build part 3, first I may want to use a U.S. map to show the data. The nutrition dataset includes different states, I think we could build a map with data when I hover over the location. Furthermore, using the ipywidgets dropdown function to let viewers select classes they are interested in, such as Obesity/Weight status, Physical Activity, and Fruits and Vegetables. When they choose a class, there is another dropdown function they could choose a related question to observe the data on the map.

1. Yuting(me): I could build a map and data on each state point.

1. Jennifer Shiyue Zhang: Build ipywidgets for selecting class, and the map will only show the class data.

1. Chung Kuang Lin: Build another dropdown function for selecting a question, connected with the first dropdown.